File: tests/leetcode/x/ruby/0269.rb
def solve(words)
  chars = words.join.chars.uniq.sort
  adj = chars.to_h { |c| [c, []] }
  indeg = chars.to_h { |c| [c, 0] }
  words.each_cons(2) do |a, b|
    m = [a.length, b.length].min
    return "" if a[0, m] == b[0, m] && a.length > b.length
    a.chars.zip(b.chars).each do |x, y|
      next if x == y
      unless adj[x].include?(y)
        adj[x] << y
        adj[x].sort!
        indeg[y] += 1
      end
      break
    end
  end
  zeros = chars.select { |c| indeg[c].zero? }.sort
  out = []
  until zeros.empty?
    c = zeros.shift
    out << c
    adj[c].each do |nei|
      indeg[nei] -= 1
      if indeg[nei].zero?
        zeros << nei
        zeros.sort!
      end
    end
  end
  out.length == chars.length ? out.join : ""
end

lines = STDIN.read.lines(chomp: true)
unless lines.empty?
  t = lines[0].to_i
  idx = 1
  out = []
  t.times do
    n = lines[idx].to_i
    idx += 1
    out << solve(lines[idx, n].map(&:strip))
    idx += n
  end
  print out.join("\n")
end

File: tests/leetcode/x/python/0269.py
import sys
from collections import deque


def solve(words):
    chars = set("".join(words))
    adj = {c: set() for c in chars}
    indeg = {c: 0 for c in chars}
    for a, b in zip(words, words[1:]):
        m = min(len(a), len(b))
        if a[:m] == b[:m] and len(a) > len(b):
            return ""
        for x, y in zip(a, b):
            if x != y:
                if y not in adj[x]:
                    adj[x].add(y)
                    indeg[y] += 1
                break
    q = deque(sorted([c for c in chars if indeg[c] == 0]))
    out = []
    while q:
        c = q.popleft()
        out.append(c)
        for nei in sorted(adj[c]):
            indeg[nei] -= 1
            if indeg[nei] == 0:
                q.append(nei)
    return "".join(out) if len(out) == len(chars) else ""


lines = sys.stdin.read().splitlines()
if lines:
    t = int(lines[0].strip())
    idx = 1
    out = []
    for _ in range(t):
        n = int(lines[idx].strip())
        idx += 1
        words = [lines[idx + i].strip() for i in range(n)]
        idx += n
        out.append(solve(words))
    sys.stdout.write("\n".join(out))
